### PR TITLE
Bugfix for Scatter and GatherElementsGrad

### PR DESCRIPTION
**Pull request by iK1D**
Current dim check for Scatter/ScatterElements and GatherElementsGrad on attribute axis is correct, it's possible that the dim value of indices on attribute axis is greater or equal to the dim value of data in the same axis. This PR is to fix this.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cpu/tensor/gather_elements_grad.cc [hunk 1]
```diff
@@ -64,7 +64,9 @@ Status GatherElementsGrad::Compute(OpKernelContext* context) const {
   }
 
   for (size_t i = 0; i < output_dims.size(); ++i) {
-    if (output_dims[i] < indices_dims[i]) {
+    // For all axes except the axis of interest, make sure that the corresponding 'indices' shape
+    // value if within bounds of the corresponding 'data' shape.
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/tensor/gather_elements_grad.cc:68]: if -> is

Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cpu/tensor/gather_elements_grad.cc:68]: this comment applies to all the files

File: onnxruntime/test/providers/cpu/tensor/scatter_op_test.cc [hunk 2]
```diff
@@ -278,5 +278,21 @@ TEST(Scatter, SameUpdateWithoutAxis) {
   scatter_same_updates_tests("ScatterElements", 11);
 }
 
+static void scatter_with_larger_indices_on_axis_tests(const char* op_name, int op_version) {
+  OpTester test(op_name, op_version);
+  test.AddAttribute<int64_t>("axis", 1);
+
+  test.AddInput<float>("data", {1, 2}, {1.0f, 2.0f});
+  test.AddInput<int64_t>("indices", {1, 4}, {0, 0, 0, 0});
+  test.AddInput<float>("updates", {1, 4}, {3.0f, 3.0f, 3.0f, 3.0f});
```
Comment (SherlockNoMad) [inline onnxruntime/test/providers/cpu/tensor/scatter_op_test.cc:287]: this is an interesting case... 
what if the updates have different values on the same entry? 
which values will got adopted?

**Review by SherlockNoMad — DISMISSED** [review]

File: onnxruntime/test/providers/cpu/tensor/scatter_op_test.cc (hunk 2 above)
Comment (iK1D) [reply onnxruntime/test/providers/cpu/tensor/scatter_op_test.cc:287]: That's why I used same values here... I don't think the spec defines this case well, so the behavior depends on the implementation.

**Review by pengwa — APPROVED** [review]
LGTM

**Pull request merged**